### PR TITLE
[None][perf] Use bf16 custom router GEMM kernel for DeepSeek-V4

### PR DESCRIPTION
**Pull request by hyukn**
Extend the existing dsv3 router GEMM custom kernel to support hidden_dim=4096 (DeepSeek-V4) in addition to 7168 (V3). For batch<=16, the custom kernel avoids cuBLAS overhead (~3us vs ~8us). For larger batches, cuBLAS fallback is preserved.

This eliminates the unnecessary fp32 cast in the router gate forward pass, yielding 3-10% end-to-end throughput improvement on B200.

@coderabbitai summary

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

To see a list of available CI bot commands, please comment `/bot help`.

### CODE REVIEW

File: tensorrt_llm/_torch/models/modeling_deepseekv4.py [hunk 1]
```diff
@@ -1325,7 +1325,9 @@ def fetch_e_score_correction_bias():
         )
 
     def forward(self, hidden_states: torch.Tensor) -> torch.Tensor:
-        return torch.nn.functional.linear(hidden_states.float(), self.weight.float())
+        return torch.ops.trtllm.dsv3_router_gemm_op(
+            hidden_states, self.weight.t(), bias=None, out_dtype=torch.float32
```
Comment (2ez4bz) [inline tensorrt_llm/_torch/models/modeling_deepseekv4.py:1329]: Just checking: is `self.weight.t()` a view or contiguous? If the former, would it be beneficial to assign the contiguous transpose to a variable that can be reused every forward?

File: tests/unittest/_torch/thop/parallel/test_dsv3_router_gemm.py [hunk 2]
```diff
@@ -10,9 +10,10 @@ def router_gemm_ref(input, weight, bias, dtype):
 @pytest.mark.parametrize(
     "num_tokens", [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16])
 @pytest.mark.parametrize("num_experts", [256])
-@pytest.mark.parametrize("hidden_size", [7168])
+@pytest.mark.parametrize("hidden_size", [7168, 4096])
```
Comment (2ez4bz) [inline tests/unittest/_torch/thop/parallel/test_dsv3_router_gemm.py:13]: Just checking: how long does this test take to run? Note that the new parameterization for `hidden_size` double the test execution time for this particular function.

**Review by 2ez4bz — APPROVED** [review]

**Review by peihu-nv — APPROVED** [review]
Tested both the accuracy and perf on B200 and GB200 MTP0/MTP1. All looking good.

File: tests/unittest/_torch/thop/parallel/test_dsv3_router_gemm.py (hunk 2 above)
Comment (hyukn) [reply tests/unittest/_torch/thop/parallel/test_dsv3_router_gemm.py:13]: I think the router GEMM should not take too long to run. I can take some quick validations.

Comment (hyukn) [reply tests/unittest/_torch/thop/parallel/test_dsv3_router_gemm.py:13]: The test takes less than 1s. It will not be a big deal. Thanks for mentioning.

File: tensorrt_llm/_torch/models/modeling_deepseekv4.py (hunk 1 above)
Comment (hyukn) [reply tensorrt_llm/_torch/models/modeling_deepseekv4.py:1329]: Thanks for the mentioning. Because the op wants the B matrix to be column major, `.t()` here will not introduce any copy. So I think cachine may not be necessary.
